[#6507][fix] Fix precision issue due to KV layout mismatch for split/concat kernels

ZhangGe6 opened the pull request:
The split/concat kernels expect KV cache in `[max_num_pages, 2, num_kv_heads, page_size, head_dim]` layout. However, the actually used KV cache layout exposed by `KVCacheManager.get_buffers()` is `[max_num_pages, 2, page_size, num_kv_heads, head_dim]`. This layout mismatch causes wrong indexing for split/concat kernels, leading to incorrect transferred prefill KV cache. This patch is a quick fix for flashinfer attn_backend.

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Option to select KV cache memory layout (NHD or HND) for attention caching.

* **Refactor**
  * Default KV cache layout changed to HND.
  * Resource manager API updated to accept and validate layout choices.

* **Chores**
  * KV buffer retrieval and reshaping adjusted to honor the chosen layout while preserving compatibility.

* **Tests**
  * Unit tests tweaked to reflect KV cache construction specifics.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

<!--
Please write the PR title by following this template:

**[JIRA ticket/NVBugs ID/GitHub issue/None][type] Summary**

Valid ticket formats:
  - JIRA ticket: [TRTLLM-1234] or [FOOBAR-123] for other FOOBAR project
  - NVBugs ID: [https://nvbugs/1234567]
  - GitHub issue: [#1234]
  - No ticket: [None]

Valid types (lowercase): [fix], [feat], [doc], [infra], [chore], etc.

Examples:
  - [TRTLLM-1234][feat] Add new feature
  - [https://nvbugs/1234567][fix] Fix some bugs
  - [#1234][doc] Update documentation
  - [None][chore] Minor clean-up
-->

## Description

<!--
Please explain the issue and the solution in short.
-->

## Test Coverage

<!--
Please list clearly what are the relevant test(s) that can safeguard the changes in the PR. This helps us to ensure we have sufficient test coverage for the PR.
-->

## GitHub Bot Help

`/bot [-h] ['run', 'kill', 'skip', 'reuse-pipeline'] ...`

Provide a user friendly way for developers to interact with a Jenkins server.

Run `/bot [-h|--help]` to print this help message.

See details below for each supported subcommand.

<details>

`run  [--reuse-test (optional)pipeline-id --disable-fail-fast --skip-test --stage-list "A10-PyTorch-1, xxx" --gpu-type "A30, H100_PCIe" --test-backend "pytorch, cpp" --add-multi-gpu-test --only-multi-gpu-test --disable-multi-gpu-test --post-merge --extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx" --detailed-log --debug(experimental)]`

Launch build/test pipelines. All previously running jobs will be killed.

`--reuse-test (optional)pipeline-id ` *(OPTIONAL)* : Allow the new pipeline to reuse build artifacts and skip successful test stages from a specified pipeline or the last pipeline if no pipeline-id is indicated. If the Git commit ID has changed, this option will be always ignored. The DEFAULT behavior of the bot is to reuse build artifacts and successful test results from the last pipeline.

`--disable-reuse-test ` *(OPTIONAL)* : Explicitly prevent the pipeline from reusing build artifacts and skipping successful test stages from a previous pipeline. Ensure that all builds and tests are run regardless of previous successes.

`--disable-fail-fast ` *(OPTIONAL)* : Disable fail fast on build/tests/infra failures.

`--skip-test ` *(OPTIONAL)* : Skip all test stages, but still run build stages, package stages and sanity check stages. Note: Does **NOT** update GitHub check status.

`--stage-list "A10-PyTorch-1, xxx"` *(OPTIONAL)* : Only run the specified test stages. Examples: "A10-PyTorch-1, xxx". Note: Does **NOT** update GitHub check status.

`--gpu-type "A30, H100_PCIe"` *(OPTIONAL)* : Only run the test stages on the specified GPU types. Examples: "A30, H100_PCIe". Note: Does **NOT** update GitHub check status.

`--test-backend "pytorch, cpp"` *(OPTIONAL)* : Skip test stages which don't match the specified backends. Only support [pytorch, cpp, tensorrt, triton]. Examples: "pytorch, cpp" (does not run test stages with tensorrt or triton backend). Note: Does **NOT** update GitHub pipeline status.

`--only-multi-gpu-test ` *(OPTIONAL)* : Only run the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--disable-multi-gpu-test ` *(OPTIONAL)* : Disable the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--add-multi-gpu-test ` *(OPTIONAL)* : Force run the multi-GPU tests in addition to running L0 pre-merge pipeline.

`--post-merge ` *(OPTIONAL)* : Run the L0 post-merge pipeline instead of the ordinary L0 pre-merge pipeline.

`--extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx"` *(OPTIONAL)* : Run the ordinary L0 pre-merge pipeline and specified test stages. Examples: --extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx".

`--detailed-log ` *(OPTIONAL)* : Enable flushing out all logs to the Jenkins console. This will significantly increase the log volume and may slow down the job.

`--debug ` *(OPTIONAL)* : **Experimental feature**. Enable access to the CI container for debugging purpose. Note: Specify exactly one stage in the `stage-list` parameter to access the appropriate container environment. Note: Does **NOT** update GitHub check status.

For guidance on mapping tests to stage names, see `docs/source/reference/ci-overview.md`
and the `scripts/test_to_stage_mapping.py` helper.

### kill

`kill  `

Kill all running builds associated with pull request.

### skip

`skip --comment COMMENT `

Skip testing for latest commit on pull request. `--comment "Reason for skipping build/test"` is required. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

### reuse-pipeline

`reuse-pipeline `

Reuse a previous pipeline to validate current commit. This action will also kill all currently running builds associated with the pull request. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

</details>

liji-nv commented:
@yuxianq to review

@ZhangGe6 Please use "git commit --amend -s" to sign off the commit. Thanks.

yuxianq commented on tensorrt_llm/_torch/attention_backend/flashinfer.py:
```diff
@@ -56,7 +56,9 @@ class FlashInferWrappers:
     workspace_buffer: Optional[torch.Tensor] = None
 
-    kv_layout: Literal["NHD", "HND"] = "NHD"
+    # cache concat/split kernels when using PD disaggregation
```
Please move this code comment to `get_buffers` to explain why we reshape to different shapes.

yuxianq commented on tensorrt_llm/_torch/pyexecutor/resource_manager.py:
```diff
@@ -615,16 +615,40 @@ def get_num_available_tokens(self, max_num_draft_tokens: int = 0) -> int:
+            kv_layout = "NHD" -> [max_num_pages, kv_factor, page_size, num_kv_heads, head_dim]
+            kv_layout = "HND" -> [max_num_pages, kv_factor, num_kv_heads, page_size, head_dim]
+
+        Note different attention backend/implementation can have different KV layouts,
```
"Note" -> "Note that"

yuxianq commented:
/bot run --disable-fail-fast

yuxianq commented:
@ZhangGe6 It seems that the last 3 commits have not been signed off, please sign off all commits to pass the DCO check, thanks~

ZhangGe6 committed:
[#6507][fix] Fix precision issue due to KV layout mismatch for split/concat kernels

The split/concat kernels expect KV cache in
"[max_num_pages, 2, num_kv_heads, page_size, head_dim]" layout. However,
the actually used KV cache layout exposed by"KVCacheManager.get_buffers()" is
"[max_num_pages, 2, page_size, num_kv_heads, head_dim]".
This layout mismatch causes wrong indexing for split/concat kernels,
leading to incorrect transferred prefill KV cache. This patch is a quick fix
for flashinfer attn_backend.

Signed-off-by: ZhangGe6 <sjtu.zg123@gmail.com>

tensorrt-cicd commented:
[PR_Github #23938](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/23938/) [ run ] triggered by Bot. Commit: `5cbcfcb`

tensorrt-cicd commented:
[PR_Github #23938](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/23938/) [ run ] completed with state `SUCCESS`. Commit: `5cbcfcb` 
[/LLM/main/L0_MergeRequest_PR pipeline #18025](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/18025/) completed with status: 'FAILURE'

yuxianq committed:
Fix CI errors.

Signed-off-by: Yuxian Qiu <142763828+yuxianq@users.noreply.github.com>

yuxianq commented:
/bot run --disable-fail-fast

yuxianq commented:
@ZhangGe6 It works. I have started to rerun CI. Thanks~

yuxianq committed:
Merge branch 'main' into fix_pd_diff_tp

tensorrt-cicd commented:
[PR_Github #24301](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/24301/) [ run ] triggered by Bot. Commit: `3c949b5`

yuxianq commented:
/bot run --disable-fail-fast
